src/tokenizedSubaccounts/PPTSA.sol

```diff
@@ -0,0 +1,559 @@
+        || (newParams.worstSpotBuyPrice < 1e18 || newParams.worstSpotBuyPrice > 1.2e18)
+        || (newParams.worstSpotSellPrice > 1e18 || newParams.worstSpotSellPrice < 0.8e18)
+        || (newParams.spotTransactionLeniency < 1e18 || newParams.spotTransactionLeniency > 1.2e18)
+        || newParams.optionVolSlippageFactor > 1e18
+        || newParams.minMarkValueToStrikeDiffRatio < newParams.maxMarkValueToStrikeDiffRatio
+        || newParams.maxMarkValueToStrikeDiffRatio > 1e20 || newParams.maxMarkValueToStrikeDiffRatio < 1e18
+        || newParams.optionMaxTimeToExpiry <= newParams.optionMinTimeToExpiry || newParams.optionMaxNegCash > 0
+        || newParams.feeFactor > 0.05e18
+    ) {
+      revert PPT_InvalidParams();
+    }
+
+    _getPPTSAStorage().ppParams = newParams;
+
+    emit PPTSAParamsSet(newParams);
+  }
+
+  ///////////////////////
+  // Action Validation //
+  ///////////////////////
+  function _verifyAction(IMatching.Action memory action, bytes32 actionHash) internal virtual override {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+
+    if (
+      action.expiry < block.timestamp + $.ppParams.minSignatureExpiry
+        || action.expiry > block.timestamp + $.ppParams.maxSignatureExpiry
+    ) {
+      revert PPT_InvalidActionExpiry();
+    }
+
+    _revokeSignature($.lastSeenHash);
+    $.lastSeenHash = actionHash;
+
+    BaseTSAAddresses memory tsaAddresses = getBaseTSAAddresses();
+
+    if (address(action.module) == address($.depositModule)) {
+      _verifyDepositAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.withdrawalModule)) {
+      _verifyWithdrawAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.rfqModule)) {
+      _verifyTradeAction(action, tsaAddresses);
+    } else {
+      revert PPT_InvalidModule();
+    }
+  }
+
+  //////////////
+  // Deposits //
+  //////////////
+
+  function _verifyDepositAction(IMatching.Action memory action, BaseTSAAddresses memory tsaAddresses) internal view {
+    IDepositModule.DepositData memory depositData = abi.decode(action.data, (IDepositModule.DepositData));
+
+    if (depositData.asset != address(tsaAddresses.wrappedDepositAsset)) {
+      revert PPT_InvalidAsset();
+    }
+
+    if (depositData.amount > tsaAddresses.depositAsset.balanceOf(address(this)) - totalPendingDeposits()) {
+      revert PPT_DepositingTooMuch();
+    }
+  }
+
+  /////////////////
+  // Withdrawals //
+  /////////////////
+
+  function _verifyWithdrawAction(IMatching.Action memory action, BaseTSAAddresses memory tsaAddresses) internal view {
+    IWithdrawalModule.WithdrawalData memory withdrawalData = abi.decode(action.data, (IWithdrawalModule.WithdrawalData));
+
+    if (withdrawalData.asset != address(tsaAddresses.wrappedDepositAsset)) {
+      revert PPT_InvalidAsset();
+    }
+
+    (int usedInTradesBalance, uint baseBalance, int cashBalance) = _getSubAccountStats();
```

Comment:
only need to check "has options open" or "exceeds max neg cash" - if not, allow any sized withdrawal through. The latter would be caught by margin requirements so it isn't even necessary to check imo

Some argument to be made that it probably shouldn't withdraw most of the balance when there is a large negative cash position; but that's not necessarily malicious so it's acceptable to not check that too closely. Logic to add some buffer is probably overkill

IMO only the first check is needed; but if you wanted to add extra safety to withdrawals, you could do something like:

```
if options open: revert

if requestedCollat > balance: revert

postWithdrawBalance = balance - requestedCollat

if cash < 0 && cash.abs() * buffer (like 5) / basePrice < postWithdrawBalance: revert
```

Reply:
Oh yes I like asserting to revert if any positions are open. 

Ill probably leave in the `requestedCollat > balance` check unless it's okay to have a negative LRT balance?

Yeah good call on the  negative cash check. ill probably leave out the buffer check to get some of the size down